src/main/kotlin/org/jitsi/nlj/transform/node/SrtpTransformerNode.kt

```diff
@@ -61,39 +62,63 @@ class SrtpTransformerNode(name: String) : MultipleOutputTransformerNode(name) {
             if (firstPacketForwardedTimestamp == -1L) {
                 firstPacketForwardedTimestamp = System.currentTimeMillis()
             }
-            val outPackets: List<PacketInfo>
-            if (!cachedPackets.isEmpty()) {
-                cachedPackets.add(packetInfo)
-                outPackets = transformList(cachedPackets)
-                cachedPackets.clear()
-            } else {
-                outPackets = if (transformer!!.transform(packetInfo))
-                    listOf(packetInfo) else emptyList()
+
+            // This is necessary in order to guarantee that only one thread accesses this.cachedPackets at a time,
+            // while requiring no synchronization in the happy case (when this.cachedPackets has been set to null).
+            var cachedPackets = this.cachedPackets
+            if (cachedPackets != null) {
+                synchronized(cachedPackets) {
+                    cachedPackets = this.cachedPackets
+                    this.cachedPackets = null
+
+                    if (cachedPackets != null) {
+                        cachedPackets!!.add(packetInfo)
+                        return transformList(cachedPackets!!)
+                    }
+                }
             }
-            return outPackets
+
+            return if (transformer!!.transform(packetInfo))
+                listOf(packetInfo) else emptyList()
```

Comment:
this code is pretty complex and we need to have it in multiple places.  given the common path will just check `cachedPackets.isEmpty()`, i suspect that using a `ConcurrentHashMap` here will be much simpler (i think nothing else would need to change from the original code other than its type) and probably no noticeable performance overhead (i'd be happy to measure that, if you like).

also, we can improve the original code here by using `it` in place of `transformer!!` since we've already used `let` to make sure it's non null (maybe we should explicitly name the `let` arg to `actualTransformer` or something to make it more obvious)

Reply:
I tried just changing it to `CopyOnWriteArrayList` (mistakenly referred to a map above) and ran the receiver harness multiple times.  CopyOnWriteArrayList doesn't do any locking on read (checking `isEmpty` in the common path so there should be no overhead there.  I ran the rtp receiver harness a few times with each:

here's a table of a few runs, tracking avg processing time (nanos) for the srtp decrypt node

|| As-is (ArrayList) | CopyOnWriteArrayList |
|---|---|---|
||21783|20826|
||19827|19931|
||21920|21947|
||20279|20608|
||21633|21787|
||19370|19694|
|avg|20802|20798|

but i'm also trying to think more about the case that's actually causing the concurrent modification exception (trying to make sure using the copy on write list will actually solve the concurrency issue correctly).  the stack trace doesn't make sense to me (SrtpTransformerNode has no line 101), was that run with some local changes you had?

Reply:
> using `it` in place of `transformer!!`

Nice



> the stack trace doesn't make sense to me (SrtpTransformerNode has no line 101), was that run with some local changes you had?

No, that's with the current code. I'm not sure what happens, but there are multiple way that it could happen. I think the most probable is: a thread starts transforming the cache and gets paused. Another thread executes stop() and empties the cache. The other options would involve retransmissions or probing packets, and I don't know if any of them are possible.

I'm happy to use `CopyOnWriteArrayList`. If you have the code already can you push it here or open another PR?